Config mapping to a Map with a List value type fails

### Describe the bug

Having a config mapping to a `Map<String, List<Double>>` fails with the `java.lang.ClassCastException`. 

Seems that the [ConfigMappingProvider#processLazyMap](https://github.com/smallrye/smallrye-config/blob/main/implementation/src/main/java/io/smallrye/config/ConfigMappingProvider.java#L416-L488) does not have a correct switch for the `CollectionProperty` and fails with class cast exception.

I can also confirm that the config mapping to `Map<String, Map<Double, Boolean>>` works for example. Again most likely because that `ConfigMappingProvider` does handle the case when value is a map.

### Expected behavior

* the map value resolves correctly in case of a `List` or `Set`

### Actual behavior

* class cast exception

### How to Reproduce?

Simple to reproduce:

```
@ConfigMapping(prefix = "some")
public interface Config {

  Map<String, List<Double>> prop();
  
}  
```

and config:

```
some:
  prop:
    value:    
      - 0.9
      - 0.99
```

leads to:

```
Caused by: java.lang.ClassCastException
	at io.smallrye.config.ConfigMappingInterface$Property.asGroup(ConfigMappingInterface.java:225)
	at io.smallrye.config.ConfigMappingProvider.processLazyMap(ConfigMappingProvider.java:481)
	at io.smallrye.config.ConfigMappingProvider.processLazyMapInGroup(ConfigMappingProvider.java:411)
	at io.smallrye.config.ConfigMappingProvider.processProperty(ConfigMappingProvider.java:215)
	at io.smallrye.config.ConfigMappingProvider.processEagerGroup(ConfigMappingProvider.java:151)
	at io.smallrye.config.ConfigMappingProvider.<init>(ConfigMappingProvider.java:80)
	at io.smallrye.config.ConfigMappingProvider$Builder.build(ConfigMappingProvider.java:955)
	at io.smallrye.config.SmallRyeConfigBuilder.build(SmallRyeConfigBuilder.java:444)
	at io.quarkus.runtime.generated.Config.readConfig(Unknown Source)
	at io.quarkus.deployment.steps.RuntimeConfigSetup.deploy(Unknown Source)
	... 13 more
```

### Output of `uname -a` or `ver`

Linux ise-Precision-5550 5.13.0-39-generic #44~20.04.1-Ubuntu SMP Thu Mar 24 16:43:35 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux

### Output of `java -version`

_No response_

### GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

2.8.0.Final

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Apache Maven 3.8.4 (9b656c72d54e5bacbed989b64718c159fe39b537) Maven home: /home/ise/.m2/wrapper/dists/apache-maven-3.8.4-bin/52ccbt68d252mdldqsfsn03jlf/apache-maven-3.8.4 Java version: 17.0.2, vendor: Private Build, runtime: /usr/lib/jvm/java-17-openjdk-amd64 Default locale: en_US, platform encoding: UTF-8 OS name: "linux", version: "5.13.0-39-generic", arch: "amd64", family: "unix"

### Additional information

_No response_

--- a/bom/application/pom.xml
+++ b/bom/application/pom.xml
@@ -40,7 +40,7 @@
         <microprofile-jwt.version>1.2</microprofile-jwt.version>
         <microprofile-lra.version>1.0</microprofile-lra.version>
         <smallrye-common.version>1.13.0</smallrye-common.version>
-        <smallrye-config.version>2.10.1</smallrye-config.version>
+        <smallrye-config.version>2.11.0</smallrye-config.version>
         <smallrye-health.version>3.2.1</smallrye-health.version>
         <smallrye-metrics.version>3.0.5</smallrye-metrics.version>
         <smallrye-open-api.version>2.1.22</smallrye-open-api.version>

--- a/integration-tests/smallrye-config/src/main/java/io/quarkus/it/smallrye/config/AppConfig.java
+++ b/integration-tests/smallrye-config/src/main/java/io/quarkus/it/smallrye/config/AppConfig.java
@@ -10,11 +10,15 @@ public interface AppConfig {
 
     Info info();
 
+    String toString();
+
     interface Info {
         @WithDefault("alias")
         String alias();
 
         @WithDefault("10")
         Integer count();
+
+        String toString();
     }
 }

--- a/integration-tests/smallrye-config/src/main/java/io/quarkus/it/smallrye/config/AppConfigResouce.java
+++ b/integration-tests/smallrye-config/src/main/java/io/quarkus/it/smallrye/config/AppConfigResouce.java
@@ -20,4 +20,10 @@ public class AppConfigResouce {
     public String getAlias() {
         return appConfig.info().alias();
     }
+
+    @GET
+    @Path("/toString")
+    public String getToString() {
+        return appConfig.toString();
+    }
 }

--- a/integration-tests/smallrye-config/src/main/java/io/quarkus/it/smallrye/config/ConfigResource.java
+++ b/integration-tests/smallrye-config/src/main/java/io/quarkus/it/smallrye/config/ConfigResource.java
@@ -23,6 +23,12 @@ public class ConfigResource {
         return Response.ok(config.getConfigValue(name)).build();
     }
 
+    @GET
+    @Path("/profiles")
+    public Response profiles() {
+        return Response.ok(config.getProfiles()).build();
+    }
+
     @RegisterForReflection(targets = {
             org.eclipse.microprofile.config.ConfigValue.class,
             io.smallrye.config.ConfigValue.class

--- a/integration-tests/smallrye-config/src/main/java/io/quarkus/it/smallrye/config/Server.java
+++ b/integration-tests/smallrye-config/src/main/java/io/quarkus/it/smallrye/config/Server.java
@@ -153,11 +153,14 @@ public interface Server extends Alias {
         Optional<List<@Size(max = 3) String>> alias();
 
         @JsonProperty
-        Map<String, Admin> admins();
+        Map<String, List<Admin>> admins();
+
+        @JsonProperty
+        Map<String, List<@Size(min = 8, max = 15) String>> firewall();
 
         interface Admin {
             @JsonProperty
-            @Size(max = 3)
+            @Size(max = 4)
             String username();
         }
     }

--- a/integration-tests/smallrye-config/src/main/resources/application.properties
+++ b/integration-tests/smallrye-config/src/main/resources/application.properties
@@ -6,9 +6,11 @@ smallrye.config.locations=config.properties
 smallrye.config.source.file.locations=src/main/smallrye-config-source-file-system
 
 # parent profile
-%prod.quarkus.config.profile.parent=common
-%test.quarkus.config.profile.parent=common
-%dev.quarkus.config.profile.parent=common
+%prod.quarkus.config.profile.parent=parent-common,common
+%test.quarkus.config.profile.parent=parent-common,common
+%dev.quarkus.config.profile.parent=parent-common,common
+
+%parent-common.parent.common.prop=parent-common
 
 %common.profile.property.shared=common
 %test.profile.property.shared=main

--- a/integration-tests/smallrye-config/src/main/resources/application.yaml
+++ b/integration-tests/smallrye-config/src/main/resources/application.yaml
@@ -74,7 +74,14 @@ cloud:
       - James
     admins:
       root:
-        username: root
+        -
+          username: root
+        -
+          username: super
+    firewall:
+      accepted:
+        - 127.0.0.1
+        - 8.8.8
 
 profile:
   main:

--- a/integration-tests/smallrye-config/src/test/java/io/quarkus/it/smallrye/config/AppConfigIT.java
+++ b/integration-tests/smallrye-config/src/test/java/io/quarkus/it/smallrye/config/AppConfigIT.java
@@ -1,0 +1,7 @@
+package io.quarkus.it.smallrye.config;
+
+import io.quarkus.test.junit.QuarkusIntegrationTest;
+
+@QuarkusIntegrationTest
+public class AppConfigIT extends AppConfigTest {
+}

--- a/integration-tests/smallrye-config/src/test/java/io/quarkus/it/smallrye/config/AppConfigTest.java
+++ b/integration-tests/smallrye-config/src/test/java/io/quarkus/it/smallrye/config/AppConfigTest.java
@@ -1,0 +1,21 @@
+package io.quarkus.it.smallrye.config;
+
+import static io.restassured.RestAssured.given;
+import static javax.ws.rs.core.Response.Status.OK;
+import static org.hamcrest.Matchers.is;
+
+import org.junit.jupiter.api.Test;
+
+import io.quarkus.test.junit.QuarkusTest;
+
+@QuarkusTest
+public class AppConfigTest {
+    @Test
+    void getToString() {
+        given()
+                .get("/app-config/toString")
+                .then()
+                .statusCode(OK.getStatusCode())
+                .body(is("AppConfig{name=app, info=Info{count=10, alias=alias}}"));
+    }
+}

--- a/integration-tests/smallrye-config/src/test/java/io/quarkus/it/smallrye/config/ServerResourceTest.java
+++ b/integration-tests/smallrye-config/src/test/java/io/quarkus/it/smallrye/config/ServerResourceTest.java
@@ -102,13 +102,14 @@ class ServerResourceTest {
         given().get("/server/validator/{prefix}", "cloud")
                 .then()
                 .statusCode(OK.getStatusCode())
-                .body("errors", hasSize(7))
+                .body("errors", hasSize(8))
                 .body("errors", hasItem("cloud.log.days must be less than or equal to 15"))
                 .body("errors", hasItem("cloud.cors.origins[1].port must be greater than or equal to 8000"))
                 .body("errors", hasItem("cloud.info.name size must be between 0 and 3"))
                 .body("errors", hasItem("cloud.info.code must be less than or equal to 3"))
                 .body("errors", hasItem("cloud.info.alias[0] size must be between 0 and 3"))
-                .body("errors", hasItem("cloud.info.admins.root.username size must be between 0 and 3"))
+                .body("errors", hasItem("cloud.info.admins.root[1].username size must be between 0 and 4"))
+                .body("errors", hasItem("cloud.info.firewall.accepted[1] size must be between 8 and 15"))
                 .body("errors", hasItem("cloud.proxy.timeout must be less than or equal to 10"));
     }
 }

--- a/integration-tests/smallrye-config/src/test/java/io/quarkus/it/smallrye/config/SmallRyeConfigTest.java
+++ b/integration-tests/smallrye-config/src/test/java/io/quarkus/it/smallrye/config/SmallRyeConfigTest.java
@@ -2,6 +2,7 @@ package io.quarkus.it.smallrye.config;
 
 import static io.restassured.RestAssured.given;
 import static javax.ws.rs.core.Response.Status.OK;
+import static org.hamcrest.Matchers.containsString;
 import static org.hamcrest.Matchers.equalTo;
 
 import org.junit.jupiter.api.Test;
@@ -97,5 +98,20 @@ public class SmallRyeConfigTest {
                 .statusCode(OK.getStatusCode())
                 .body("value", equalTo("1234"))
                 .body("sourceName", equalTo("UserConfigSource"));
+    }
+
+    @Test
+    void profiles() {
+        given()
+                .get("/config/profiles")
+                .then()
+                .statusCode(OK.getStatusCode())
+                .body(containsString("parent-common"));
+
+        given()
+                .get("/config/{name}", "parent.common.prop")
+                .then()
+                .statusCode(OK.getStatusCode())
+                .body("value", equalTo("parent-common"));
     }
 }

